### ISSUE TITLE
Error: ipfs repo needs migration

### ISSUE THREAD

**Issue by davidak** [NixOS/nixpkgs#113343]
**Describe the bug**
After updating the channel, IPFS does not start anymore.

```
Feb 16 22:35:29 nas systemd[1]: Starting IPFS Initializer...
Feb 16 22:35:29 nas ipfs-init-start[1076]: Error: ipfs repo needs migration
Feb 16 22:35:29 nas systemd[1]: ipfs-init.service: Main process exited, code=exited, status=1/FAILURE
Feb 16 22:35:29 nas systemd[1]: ipfs-init.service: Failed with result 'exit-code'.
Feb 16 22:35:29 nas systemd[1]: Failed to start IPFS Initializer.
```

It would be nice if a migration would be done automatically if needed.

Or at least document the process in the release notes.

**To Reproduce**
1. ...
2. ...
3. ...

**Expected behavior**
IPFS just working

**Notify maintainers**
@fpletz

**Metadata**
 - system: `"x86_64-linux"`
 - host os: `Linux 5.4.97, NixOS, 20.09.3124.2118cf551b9 (Nightingale)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.10`
 - channels(root): `"nixos-20.09.3124.2118cf551b9, nixos-unstable-21.05pre269929.ff96a0fa563"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute: ipfs
# a list of nixos modules affected by the problem
module: ipfs
```

**Comment by davidak** [comment]
## Manual Migration

```
[root@nas:~]# nix run nixpkgs.ipfs-migrator -c sudo -u ipfs IPFS_PATH=/var/lib/ipfs/ fs-repo-migrations
Found fs-repo version 7 at /var/lib/ipfs/
Do you want to upgrade this to version 10? [y/n] y
===> Running migration 7 to 8...
applying 7-to-8 repo migration
locking repo at "/var/lib/ipfs/"
  - verifying version is '7'
> Upgrading config to new format
updated version file
===> Migration 7 to 8 succeeded!
===> Running migration 8 to 9...
applying 8-to-9 repo migration
updated version file
===> Migration 8 to 9 succeeded!
===> Running migration 9 to 10...
applying 9-to-10 repo migration
locking repo at "/var/lib/ipfs/"
  - verifying version is '9'
> Upgrading config to new format
updated version file
===> Migration 9 to 10 succeeded!
```

When i run it again, nothing happens.

```
[root@nas:~]# nix run nixpkgs.ipfs-migrator -c sudo -u ipfs IPFS_PATH=/var/lib/ipfs/ fs-repo-migrations
ipfs migration: already at target version number
```

So we might just put it in a systemd service that runs every time before IPFS starts.

But it might break and lead to data loss. That would be very bad! It is advised to backup before doing this.

So it has to be discussed if we do such dangerous operations on peoples data automatically or just document how to do it manually.

**Comment by aanderse** [comment]
There are strong opinions both for automatic migrations and against. Some software will bake migrations right into their primary executable and not even give you a chance. Other software will **not** fail if you don't run a migration and silently introduce strange breaking behavior that can be difficult to diagnose. Some migrations can be awkward to run on NixOS. As you mentioned some migrations are not safe to run without human interaction. Some migration programs aren't idempotent. Some software can damage it's own state if upgraded and run before migrations are executed. And to top it all off, the NixOS module system doesn't give us the ability to easily prevent a service from starting the first time after an upgrade.

In summary this topic is very complex and hasn't been discussed in any public forum in detail. It would be nice if NixOS had a good solution for this in general and promoted standard desired behavior to module authors.

**Comment by kevincox** [comment]
I think we should run the migrations automatically. It isn't great to have your system stop working.

They don't really *recommend* backups, they say to do them if you want to be extra safe. I think we should just rely on the user having their own backups anyways. https://github.com/ipfs/fs-repo-migrations/blob/master/run.md#step-0-back-up-your-repo-optional

We should probably have a flag to control this, but I think default-enabled would be fine.

**Comment by aanderse** [comment]
ping @Infinisil because they have become increasingly interested in these types of topics lately.
ping @grahamc who wants to smash all things state.

:smile:

**Comment by ThibaultLemaire** [comment]
@davidak 

``` fish
❯ nix run nixpkgs.ipfs-migrator -c sudo -u ipfs IPFS_PATH=/var/lib/ipfs/ fs-repo-migrations
error: unrecognised flag '-c'
Try 'nix --help' for more information.
```

What am I doing wrong?

**Comment by Luflosi** [comment]
I used
```sh
nix shell nixpkgs#ipfs-migrator --command sudo -u ipfs IPFS_PATH=/path/to/ipfs/repo fs-repo-migrations
```

**Comment by Ericson2314** [comment]
Is there least some sort of one-shot systemd job that we could create that users would run by hand instead of the above? How are e.g. postgresql migrations handled today?

**Comment by happysalada** [comment]
AFAIK, on postgresql migrations are always handled by applications themselves. Each application defines it's own ExecPreStart to run it's own migration.

I was thinking that we could have a separate one-shot systemd job, and an option to run migrations automatically. If the automated migrations option is checked, then ipfs service would `want` the migration job, and if it's not checked, then the user can run manually that job. I'm having a look at running IPFS at the moment. I've got a [PR with one small fix open](https://github.com/NixOS/nixpkgs/pull/135850) in case anyone is interested.

**Comment by kevincox** [comment]
> Is there least some sort of one-shot systemd job that we could create that users would run by hand instead of the above? How are e.g. postgresql migrations handled today?

You can definitely run fs-repo-migrations in a dependency or PreStart command.

> AFAIK, on postgresql migrations are always handled by applications themselves.

Are you referring to Postgres version migrations or application schema migrations? Applications migrating their own schema is definitely the common case but I have yet to see an application that will automatically migrate your postgres from 9.2 to 9.3. The IPFS case is somewhat in the middle because it is both the application and the database. But in this case I think it makes sense to act like the application and run the migrations on startup.

**Comment by happysalada** [comment]
yes, I was referring to schema migration. I've never seen version migration either.

**Comment by happysalada** [comment]
I went with the easy solution, just running the migrator on preStart.
https://github.com/NixOS/nixpkgs/pull/136261
The repo migration itself said that those were safe https://github.com/ipfs/fs-repo-migrations/blob/master/run.md
Happy to take feedback.
